### DISCUSSION TITLE
MediaPicker.CapturePhotoAsync not returning control to caller

### DISCUSSION THREAD

**Discussion by jamesbgibson** [dotnet/maui#21956] [General]
I'm having a puzzling issue with MediaPicker.CapturePhotoAsync on the iOS platform.  The PickPhotoAsync method works as expected, but the CapturePhotoAsync presents this weird problem: The iPhone camera form is presented, I take the picture, I press the 'Use Photo' button, the camera form closes and the calling form is redisplayed.  However, the calling page never receives control back, leaving that page in a hung state.  Searching through this form for answers relating to MediaPicker doesn't yield much, leading me to at first assume something is wrong with the way I've coded this, but all my attempts since at finding a solution have failed.  This is occurring in the context of upgrading a Xamarin app that used the Essential MediaPicker class.

**Comment by prabhav-mehra** [comment]
please share the code you have in place right now

**Reply by jamesbgibson** [reply]
Here's the method where this condition occurs:

        public async System.Threading.Tasks.Task<int> OnClicked_TakePhoto(object sender, System.EventArgs eventargs)
        {
            System.Byte[] bytes;
            System.DateTime datetimeNow = System.DateTime.Now;
            System.DateTime datetimeNowUTC = System.DateTime.UtcNow;
            System.IO.Stream stream;
            System.IO.MemoryStream memorystream;
            ExtMobi_ActivityBar xmractivitybar = null;

            if (!ExtMobi_Global.IsNetworkConnected())
            {
                await base.DisplayConnectionAlert();
                return 0;
            }

            xmractivitybar = new ExtMobi_ActivityBar();
            xmractivitybar.Prepare();
            xmractivitybar.Status = "Accessing camera...";
            xmractivitybar.ActivityIndicator.IsRunning = true;
            base.Body.Children.Insert(0, xmractivitybar);

            try
            {
                if (!MediaPicker.Default.IsCaptureSupported)
                {
                    await ExtMobi_Util.Alert("Unable to proceed -- the camera is not available");
                    return 1;
                }

                PermissionStatus status = await Permissions.RequestAsync<Permissions.Photos>();
                if (status != PermissionStatus.Granted)
                {
                    await ExtMobi_Util.Alert("Unable to proceed -- taking a photo was not authorized.");
                    return 1;
                }

                var options = new MediaPickerOptions();

                var fileResult = await MediaPicker.Default.CapturePhotoAsync(options);
                if (fileResult == null)
                {
                    Console.WriteLine(ExtCore_Debug.GetSourceLoc());
                    await base.DisplayAlert("Debug", "fileResult == null", "Close");
                    return 0;
                }

                Console.WriteLine(ExtCore_Debug.GetSourceLoc());
                await base.DisplayAlert("Debug", "Photo taken", "Close");

                stream = File.OpenRead(fileResult.FullPath);

                memorystream = new System.IO.MemoryStream(10000000);
                stream.CopyTo(memorystream);
                bytes = memorystream.ToArray();

                xmrgalleryphotoedit = new TrlMobi_GalleryPhotoEdit();
                xmrgalleryphotoedit.CreateNew = true;
                xmrgalleryphotoedit.bytesImage = bytes;
                xmrgalleryphotoedit.sqldocumentreferenceFolder = sqldocumentreferenceOwner;
                xmrgalleryphotoedit.rcdDocumentReferenceFolder = TreeRecordOwner;
                xmrgalleryphotoedit.OwnerId = OwnerId;
                xmrgalleryphotoedit.ImageType = "jpg";
                xmrgalleryphotoedit.InitialTitle = "Create Photo";
                xmrgalleryphotoedit.GalleryPage = this;
                await xmrgalleryphotoedit.PrepareControls();
                await base.PushAsync(xmrgalleryphotoedit);
            }
            finally
            {
                if (xmractivitybar != null)
                {
                    xmractivitybar.ActivityIndicator.IsRunning = false;
                    base.Body.Children.Remove(xmractivitybar);
                }
            }

            return 0;
        }

**Reply by thiemf** [reply]
We have a similar issue that an alert doesn't show up when displayed after `CapturePhotoAsync`. 
Can see the message in the logs:
[Presentation] Attempt to present <UIAlertController: 0x7fb70eb5aa00> on <UIImagePickerController: 0x7fb70da41a00> (from <UIImagePickerController: 0x7fb70da41a00>) whose view is not in the window hierarchy.
I'm thinking the issue is that `CapturePhotoAsync` returns too early, i.e. while the UIImagePickerController is still being shown or while the dismiss animation hasn't completed yet. 
I added a one second daley after the call to `CapturePhotoAsync` in order to wait for the dismiss animation to complete. 
Would be better if `CapturePhotoAsync` could actually wait for the picker to be dismissed before returning the result.

**Reply by jamesbgibson** [reply]
Thank you, thiemf.  To circumvent this issue, we temporarily switched to using James Montemagno's Plugin.Media code.  The preferred solution of course is to use the Maui components.  When we next revisit this issue, I'll give your suggestion a try.
